README.md
# Meme Generation in R

[![Build Status](https://travis-ci.org/leeper/meme.png?branch=master)](https://travis-ci.org/leeper/meme)

According to [Wikipedia](http://en.wikipedia.org/wiki/Internet_meme):

>An Internet meme (/'mi?m/ meem) is an activity, concept, catchphrase or piece of media which spreads, often as mimicry, from person to person via the Internet.

**meme** is an R package that connects with several online image generators to create internet memes, which can then be loaded into R, plotted, and saved locally.

## Installation ##

At some point **meme** will be [available on CRAN](http://cran.r-project.org/web/packages/meme/index.html) and installed from within R from your favorite CRAN mirror:

```
install.packages("meme")
```

The latest development version, available here, can be installed directly using Hadley Wickham's [devtools](http://cran.r-project.org/web/packages/devtools/index.html) package:

```
if(!require("devtools")) install.packages("devtools")
library("devtools")
install_github("leeper/meme")
```

## Examples ##

Here are some simple examples using the package and the default [memecaptain](http://memecaptain.com/) generator. To get started, you can grab some image templates from one of the meme generator sites. We'll use memecaptain because it doesn't require a username and password.


```r
library("meme")
templates <- get_templates("memecaptain")
```

You can use the `plot.meme_template` method to plot all templates from a given site to help you figure out which one you want to use. Here's some code to do it:

```
par( mfrow = n2mfrow(length(templates)), mar=rep(0,4), mgp=rep(0,3))
invisible(lapply(templates, plot))
```

Here are some example memes. These could probably be better, so pull requests are welcome!


```r
plot(allthethings <- create_meme(templates[[2]], "Code", "all the things!"))
```

![plot of chunk allthethings](inst/figures/allthethings.png) 


```r
plot(simply <- create_meme(templates[[10]], "One does not simply", "reshape data from wide to long"))
```

![plot of chunk simply](inst/figures/simply.png) 


```r
plot(yuno <- create_meme(templates[[55]], "R CMD check", "y u no work?"))
```

![plot of chunk yuno](inst/figures/yuno.png) 


```r
u <- "http://img4.wikia.nocookie.net/__cb20131009153844/villains/images/f/f6/Grumpy_Cat.jpg"
plot(grumpy <- create_meme(u, "", "R-help"))
```

```
## 'template' is not of class 'meme_template'. Template assumed to be an image URL and meme will be generated with site 'memecaptain'.
```

![plot of chunk grumpy](inst/figures/grumpy.png) 



NEWS
# CHANGES TO meme 0.1 #

* First version released


R/get_memes.r
get_memes <- function(site = "memegenerator", type = NULL, query = NULL, ...){
    if(site == "imgflip") {
        stop("'imgflip' not currently supported")
    } else if(site == "memegenerator") {
        base <- 'http://version1.api.memegenerator.net/'
        if(is.null(type))
            type <- 'popular'
        if(type == 'new') {
            u <- paste0(base,'Instances_Select_ByNew?pageIndex=0&pageSize=24')
        } else if(type == 'popular') {
            u <- paste0(base,'Instances_Select_ByPopular?pageIndex=0&pageSize=24&days=7')
        } else if(type == 'trending') {
            u <- paste0(base,'Instances_Select_ByTopics')
        } else if(type == 'topic') {
            u <- paste0(base,'Instances_Select_ByTopic')
        } else {
            stop("Specified 'type' not recognized")
        }
    } else if(site == "memecaptain") {
        stop("'memecaptain' not currently supported")
    } else if(site == "imgur") {
        u <- "https://api.imgur.com/3/memegen/defaults"
        stop("'imgur' not currently supported")
    } else {
        stop("Only 'memegenerator' is currently supported!")
    }
    b <- basicTextGatherer()
    h <- basicHeaderGatherer()
    curlPerform(url = u, writefunction = b$update, headerfunction = h$update,
                ssl.verifypeer = 1L, ssl.verifyhost = 2L,
                cainfo = system.file("CurlSSL", "cacert.pem", package = "RCurl"), ...)
    if(h$value()["status"] != "200")
        warning("HTTP Status", h$value()["status"], ": ", h$value()["statusMessage"])
    out <- fromJSON(b$value())
    if(site == "imgflip") {
        return(NULL)
    } else if(site == 'memegenerator'){
        if(out$success)
            return(lapply(out$result, structure, class = "meme", site = site))
        else
            return(NULL)
    } else if(site == "memecaptain") {
        return(NULL)
    } else {
        return(NULL)
    }
}


R/create_meme.r
create_meme <-
function(template, upper = NULL, lower = NULL, username = NULL, password = NULL, site = NULL, 
         font = NULL, language = NULL, ...){
    if(class(template) != 'meme_template' & class(template) == 'character') {
        message("'template' is not of class 'meme_template'. Template assumed to be an image URL and meme will be generated with site 'memecaptain'.")
        template <- list(url = template)
        site <- 'memecaptain'
    } else {
        site <- attr(template, 'site')
    }
    if(site == "imgflip") {
        base <- "https://api.imgflip.com/caption_image?"
        u <- paste0(base, 'template_id=', template$id,
                    '&username=', username, '&password=', password,
                    '&text0=',curlEscape(upper), '&text1=', curlEscape(lower),
                    if(!is.null(font)) paste('&font=', font))
    } else if(site == "memegenerator") {
        base <- 'https://version1.api.memegenerator.net/Instance_Create?'
        u <- paste0(base, 'generatorID=', template$generatorID,
                    '&username=', username, '&password=', password,
                    '&text0=',curlEscape(upper), '&text1=', curlEscape(lower),
                    if(!is.null(language)) paste('&languageCode=', language))
    } else if(site == "memecaptain") {
        u <- paste0("http://v1.memecaptain.com/i?",'u=',curlEscape(template$url),'&t1=',curlEscape(upper),'&t2=',curlEscape(lower))
    } else {
        stop("Only 'imgflip', 'memegenerator', and 'memecaptain' are currently supported!")
    }
    
    if(site %in% c('imgflip', 'memegenerator')){
        b <- dynCurlReader()
        h <- basicHeaderGatherer()
        curlPerform(url = u, writefunction = b$update, headerfunction = h$update,
                    ssl.verifypeer = 1L, ssl.verifyhost = 2L,
                    cainfo = system.file("CurlSSL", "cacert.pem", package = "RCurl"), ...)
        if(h$value()["status"] != "200")
            warning("HTTP Status ", h$value()["status"], ": ", h$value()["statusMessage"])
        
        if(site == "imgflip") {
            out <- fromJSON(b$value())
            if(out$success)
                return(structure(c(as.list(out$data), meme = getBinaryURL(out$data["url"])), 
                                 class = "meme", site = site))
            else
                stop(out$error_message)
        } else if(site == 'memegenerator'){
            out <- fromJSON(b$value())
            if(out$success)
                return(structure(out, class = "meme", site = site))
            else
                return(NULL)
        }
    } else if(site == "memecaptain") {
        h <- basicHeaderGatherer()
        out <- getBinaryURL(u, headerfunction = h$update)
        if(h$value()["status"] != "200")
            warning("HTTP Status ", h$value()["status"], ": ", h$value()["statusMessage"])
        return(structure(list(url = u, meme = out), class = "meme", site = site))
    } else {
        return(NULL)
    }
    
}

plot.meme <- function(x, ...){
    j <- .readimage(x)
    plot(NULL, xlim=c(0,1), ylim=c(0,1), xaxt='n', yaxt='n', 
         xlab = "", ylab = "", main = "",
         xaxs='i', yaxs='i', mar=rep(0,4), mgp=rep(0,3), ...)
    rasterImage(j, 0,0,1,1)
    invisible(x)
}


R/get_templates.r
get_templates <- function(site = "memecaptain", type = NULL, query = NULL, ...){
    if(site == "imgflip") {
        u <- "https://api.imgflip.com/get_memes"
    } else if(site == "memegenerator") {
        base <- 'http://version1.api.memegenerator.net/'
        if(is.null(type))
            type <- 'popular'
        if(type == 'new') {
            u <- paste0(base,'Generators_Select_ByNew?pageIndex=0&pageSize=24')
        } else if(type == 'popular') {
            u <- paste0(base,'Generators_Select_ByPopular?pageIndex=0&pageSize=24&days=7')
        } else if(type == 'trending') {
            u <- paste0(base,'Generators_Select_ByTrending')
        } else if(type == 'related') {
            u <- paste0(base,'Generators_Select_Related_ByDisplayName?displayName=', curlEscape(query))
        } else if(type == 'search') {
            u <- paste0(base,'Generators_Select_Related_ByDisplayName?pageIndex=0&pageSize=24&q=', curlEscape(query))
        } else {
            stop("Specified 'type' not recognized")
        }
    } else if(site == "memecaptain") {
        u <- "http://v1.memecaptain.com/source_images.json"
    } else if(site == "imgur") {
        u <- "https://api.imgur.com/3/memegen/defaults"
        stop("'imgur' not currently supported")
    } else {
        stop("Only 'imgflip', 'memegenerator', and 'memecaptain' are currently supported!")
    }
    b <- basicTextGatherer()
    h <- basicHeaderGatherer()
    curlPerform(url = u, writefunction = b$update, headerfunction = h$update,
                ssl.verifypeer = 1L, ssl.verifyhost = 2L,
                cainfo = system.file("CurlSSL", "cacert.pem", package = "RCurl"), ...)
    if(h$value()["status"] != "200")
        warning("HTTP Status", h$value()["status"], ": ", h$value()["statusMessage"])
    out <- fromJSON(b$value())
    if(site == "imgflip") {
        if(out$success)
            return(lapply(out$data$memes, structure, class = "meme_template", site = site))
        else
            stop(out$error_message)
    } else if(site == 'memegenerator'){
        if(out$success)
            return(lapply(out$result, structure, class = "meme_template", site = site))
        else
            return(NULL)
    } else if(site == "memecaptain") {
        return(lapply(out$images, structure, class = "meme_template", site = site))
    } else {
        return(NULL)
    }
}

plot.meme_template <- function(x, ...){
    j <- .readimage(x)
    plot(NULL, xlim=c(0,1), ylim=c(0,1), xaxt='n', yaxt='n',
         xlab = "", ylab = "", main = "",
         xaxs='i', yaxs='i', mar=rep(0,4), mgp=rep(0,3), ...)
    rasterImage(j, 0,0,1,1)
    invisible(x)
}


R/meme.r
# imgflip
# api docs: https://api.imgflip.com/
# signup: https://imgflip.com/signup

# get_memes
# https://api.imgflip.com/get_memes


get_templates <- function(site = "imgflip", type = NULL, query = NULL, ...){
    if(site == "imgflip") {
        u <- "https://api.imgflip.com/get_memes"
    } else if(site == "memegenerator") {
        base <- 'http://version1.api.memegenerator.net/'
        if(is.null(type))
            type <- 'popular'
        if(type == 'new') {
            u <- paste0(base,'Generators_Select_ByNew?pageIndex=0&pageSize=24')
        } else if(type == 'popular') {
            u <- paste0(base,'Generators_Select_ByPopular?pageIndex=0&pageSize=24&days=7')
        } else if(type == 'trending') {
            u <- paste0(base,'Generators_Select_ByTrending')
        } else if(type == 'name') {
            u <- paste0(base,'Generators_Select_Related_ByDisplayName?displayName=', curlEscape(query))
        } else if(type == 'related') {
            u <- paste0(base,'Generators_Select_Related_ByDisplayName?displayName=', curlEscape(query))
        } else if(type == 'search') {
            u <- paste0(base,'Generators_Select_Related_ByDisplayName?pageIndex=0&pageSize=24&q=', curlEscape(query))
        else {
            stop("Specified 'type' not recognized")
        }
    } else if(site == "memecaptain") {
        u <- "http://v1.memecaptain.com/source_images.json"
    } else if(site == "imgur") {
        u <- "https://api.imgur.com/3/memegen/defaults"
        stop("'imgur' not currently supported")
    } else {
        stop("Only 'imgflip', 'memegenerator', and 'memecaptain' are currently supported!")
    }
    b <- basicTextGatherer()
    h <- basicHeaderGatherer()
    curlPerform(url = u, writefunction = b$update, headerfunction = h$update,
                ssl.verifypeer = 1L, ssl.verifyhost = 2L,
                cainfo = system.file("CurlSSL", "cacert.pem", package = "RCurl"), ...)
    if(h$value()["status"] != "200")
        warning("HTTP Status", h$value()["status"], ": ", h$value()["statusMessage"])
    out <- fromJSON(b$value())
    if(site == "imgflip") {
        if(out$success)
            return(lapply(out$data$memes, structure, class = "meme_template", site = site))
        else
            stop(out$error_message)
    } else if(site == 'memegenerator'){
        if(out$success)
            return(lapply(out$result, structure, class = "meme_template", site = site))
        else
            return(NULL)
    } else if(site == "memecaptain") {
        return(lapply(out$images, structure, class = "meme_template", site = site))
    } else {
        return(NULL)
    }
}

create_meme <-
function(template, upper, lower, username, password, site = NULL, 
         font = NULL, language = NULL, ...){
    if(class(template) != 'meme_template' & class(template == 'character')) {
        warning("'template' is not of class 'meme_template'. Template assumted to be an image URL and meme will be generated with site 'memecaptain'.")
        site <- 'memecaptain'
    }
    if(site == "imgflip") {
        base <- "https://api.imgflip.com/caption_image?"
        u <- paste(base, 'template_id=', template,
                   '&username=', username, '&password=', password,
                   '&text0=',upper, '&text1=', lower,
                   if(!is.null(font)) paste('&font=', font))
    } else if(site == "memegenerator") {
        base <- 'https://version1.api.memegenerator.net/Instance_Create?'
        u <- paste(base, 'generatorID=', template,
                   '&username=', username, '&password=', password,
                   '&text0=',upper, '&text1=', lower,
                   if(!is.null(language)) paste('&languageCode=', language))
    } else if(site == "memecaptain") {
        u <- paste0("http://v1.memecaptain.com/i?",'u=',template,'&t1=',upper,'&t2=',lower)
    } else {
        stop("Only 'imgflip', 'memegenerator', and 'memecaptain' are currently supported!")
    }
    
    if(site %in% c('imgflip', 'memegenerator'){
        b <- dynCurlReader()
        h <- basicHeaderGatherer()
        curlPerform(url = u, writefunction = b$update, headerfunction = h$update,
                    ssl.verifypeer = 1L, ssl.verifyhost = 2L,
                    cainfo = system.file("CurlSSL", "cacert.pem", package = "RCurl"), ...)
        if(h$value()["status"] != "200")
            warning("HTTP Status", h$value()["status"], ": ", h$value()["statusMessage"])
        
        if(site == "imgflip") {
            out <- fromJSON(b$value())
            if(out$success)
                return(structure(c(as.list(out$data), meme = getBinaryURL(out$data$url)), 
                                 class = "meme", site = site))
            else
                stop(out$error_message)
        } else if(site == 'memegenerator'){
            out <- fromJSON(b$value())
            if(out$success)
                return(structure(out, class = "meme", site = site))
            else
                return(NULL)
        }
    } else if(site == "memecaptain") {
        out <- getBinaryURL(u)
        return(structure(list(url = u, meme = out), class = "meme", site = site))
    } else {
        return(NULL)
    }
    
}

plot.meme_template <- function(x, ...){
    if(attr(x,'site') %in% c('imgflip','memecaptain'))
        u <- x$url
    else if(attr(x,'site') == 'memegenerator')
        u <- x$imageUrl
    j <- readJPEG(u)
    rasterImage(j, ...)
    invisible(x)
}

plot.meme <- function(x, ...){
    if('meme' %in% names(x)) {
        j <- readJPEG(x$meme)
    } else {
        if(attr(x,'site') == 'imgflip')
            j <- readJPEG(getBinaryURL(x$url))
        else if(attr(x,'site') == 'memegenerator')
            j <- readJPEG(x$imageUrl)
        else if(attr(x,'site') == 'memecaptain')
            stop("If site=='memecaptain', x must supply 'meme' field")
    }
    par(mar=rep(0,4), mgp=rep(0,3))
    plot(NULL, xlim=c(0,1), ylim=c(0,1), xaxs='i', yaxs='i')
    rasterImage(j, 0,0,1,1)
    invisible(x)
}

get_memes <- function(site = 'imgflip'){
    if(site == "imgflip") {
        stop("'imgflip' not currently supported")
    } else if(site == "memegenerator") {
        
    }
}
